proto/kv-connect.md

```diff
@@ -0,0 +1,448 @@
+that the client is attempting to connect to.
+
+If the access token is invalid, no overlapping protocol versions were found, the
+request body is invalid, or another error is encountered that prevent the server
+from returning database metadata, the server MUST respond with a 4xx class HTTP
+status response and a plain text body containing a human readable error message.
+The server MAY return a 3xx class HTTP redirect response. In this case, the
+client MUST follow the redirect and perform the metadata exchange with the new
+URL as the base URL.
+
+The server MAY return a 5xx class HTTP status response at any time if the server
+is unable to process the request right now due to a temporary condition. The
+server SHOULD include a plain text body with a human readable error message in
+this case.
+
+If the access token is valid and a database is found, the server MUST respond
+with a 200 OK response. This response MUST include a `Content-Type` header with
+the value `application/json`. The text encoding of the response body MUST be
+UTF-8. The response body MUST be a JSON object that adheres to the the JSON
+schema defined in the
+`schema/kv-metadata-exchange-response.v<protocolversion>.json` file.
+
+If the client fails to receive a response from the server due to a network
+error, or a 5xx class HTTP status, the client SHOULD retry the request using an
+exponential backoff strategy with unlimited retries.
+
+The client MUST verify that the response is a 200 OK, 3xx class redirect
+response, or 4xx class HTTP status response. If the response is a 3xx class
+redirect response, the client MUST follow the redirect and perform the metadata
+exchange with the new URL as the base URL. If the response is a 4xx class HTTP
+status response, the client MUST fatally abort the metadata exchange and display
+the error message to the user. If the response is a 200 OK response, the client
+MUST verify that the response body adheres to the JSON schema defined in the
+`schema/kv-metadata-exchange-response.v<protocolversion>.json` file. If the
+response body is invalid, the client MUST fatally abort the metadata exchange
+and display the error message to the user.
+
+If the response body is valid, the client MUST verify that the `version`
+property is a protocol version that the client supports. If the `version` is not
+a protocol version that the client supports, the client MUST fatally abort the
+metadata exchange and display an error message to the user.
+
+If client MUST verify that at least one of the entries in the `endpoints` array
+has a `consistency` property that is equal to `strong`. If no such entry is
+found, the client MUST fatally abort the metadata exchange and display an error
+message to the user.
+
+If the client is using protocol version 2 or higher, the client MUST parse each
+URL in the endpoints array and resolve it against the base URL of the metadata
+exchange request. If the client followed a redirect during the metadata exchange
```

Comment:
For my education, what's changing in v2 that requires the URL parsing?

Reply:
The `url` in the `endpoints` array can now be a URL relative to the metadata URL.

```jsonc
// Hosted at https://example.com/metadata/1234
{
  "version": 2,
  "endpoints": [
    {
      "url": "/data/v1", // resolved to https://example.com/data/v1
      "consistency": "strong"
    }
  ],
  // ...
}
```